src/client/connect/dns.rs

```diff
@@ -333,4 +384,34 @@ mod tests {
 
         assert_eq!(addrs.next(), Some(expected));
     }
+
+    #[test]
+    fn ipv6_addr_with_zone_try_parse() {
+        let uri = ::http::Uri::from_static("http://[fe80::1:2:3:4%25eth0]:8080/");
+        let dst = super::super::Destination { uri };
+
+        let addr = Ipv6AddrWithZone::try_parse(
+            dst.host(),
+            dst.port().expect("port")
+        ).expect("try_parse");
+
+        assert_eq!(addr.as_str, "fe80::1:2:3:4%eth0");
+
+        let expected = "[fe80::1:2:3:4]:8080".parse::<SocketAddrV6>().expect("expected");
+
+        assert_eq!(addr.partial_addr, expected);
+    }
+
+    #[test]
+    fn ipv6_addr_with_zone_try_parse_not_link_local() {
+        let uri = ::http::Uri::from_static("http://[::1%25eth0]:8080/");
+        let dst = super::super::Destination { uri };
+
+        let addr = Ipv6AddrWithZone::try_parse(
+            dst.host(),
+            dst.port().expect("port")
+        );
+
+        assert_eq!(addr.map(|a| format!("{:?}", a)), None);
+    }
```

Comment:
We'll need a test to make sure clients don't pass the scope to servers.

Reply:
Do you mean a test in this PR that hyper remembers to strip it somewhere?

Reply:
Yeah, at least in the HOST header (not sure if we need to do this elsewhere). Specifically in the HOST header (but anywhere else we might be sending the URL to a remote server). Otherwise, this could potentially open up some attack surfaces. I think [rfc7404](https://tools.ietf.org/html/rfc7404) speaks more about this.

Reply:
Sorry, wrong rfc. The right one is [rfc6874](https://tools.ietf.org/html/rfc6874), which states in section 4:

> An HTTP client, proxy, or other intermediary MUST remove any ZoneID attached to an outgoing URI, as it has only local significance at the sending host.